--- a/src/scanner.c
+++ b/src/scanner.c
@@ -125,6 +125,7 @@
     }
 
     if (state->start_tag != NULL && strcmp(state->start_tag, start_tag) == 0) {
+      free(start_tag);
       return false;
     }
 
@@ -170,19 +171,21 @@
   }
 
   memcpy(buffer, state->start_tag, tag_length);
-  if (state->start_tag != NULL) {
-    free(state->start_tag);
-    state->start_tag = NULL;
-  }
   return tag_length;
 }
 
 void tree_sitter_sql_external_scanner_deserialize(void *payload, const char *buffer, unsigned length) {
   LexerState *state = (LexerState *)payload;
-  state->start_tag = NULL;
+  if (state->start_tag != NULL) {
+    free(state->start_tag);
+    state->start_tag = NULL;
+  }
   // A length of 1 can't exists.
   if (length > 1) {
     state->start_tag = malloc(length);
+    if (state->start_tag == NULL) {
+      return;
+    }
     memcpy(state->start_tag, buffer, length);
   }
 }
